integration_tests/test_domestic_migration.py

```diff
@@ -114,43 +123,126 @@ def test_individuals_move_into_existing_households(simulants_on_adjacent_timeste
+            before[household_movers]["household_details.housing_type"] == "Standard"
+        ).all()
 
 
-def test_only_living_people_move(simulants_on_adjacent_timesteps):
+def test_only_living_people_change_households(simulants_on_adjacent_timesteps):
     for before, after in simulants_on_adjacent_timesteps:
-        movers = before["address_id"] != after["address_id"]
+        movers = before["household_id"] != after["household_id"]
         assert after[movers]["tracked"].all()
         assert (after[movers]["alive"] == "alive").all()
+
+
+@pytest.mark.parametrize(
+    "unit_id_col, address_id_col",
+    [
+        ("employer_id", "business_details.employer_address_id"),
+        ("household_id", "household_details.address_id"),
+    ],
+)
+def test_unit_address_uniqueness(populations, unit_id_col, address_id_col):
+    """Check that all units (households and businesses) have unique details and
+    also that simulants of the same unit have the same details
+    """
+    address_cols = [unit_id_col, address_id_col]  # TODO: add state/puma
+    for pop in populations:
+        # Check that all simulants in the same unit have the same address
+        assert (pop.groupby(unit_id_col)[address_cols].nunique() == 1).all().all()
+        # Check that all units have unique addresses
+        assert len(pop[address_cols].drop_duplicates()) == len(pop[unit_id_col].unique())
+
+    # # TODO implement state/puma checks
+    # # Ensure pumas map to correct state. This will be an imperfect test
+    # # because pumas are only unique within states and so one puma can
+    # # exist in multiple states. Here we only ensure no impossible pumas
+    # # exist in each state
+    # state_puma_map = get_state_puma_map(
+    #     sim._data.artifact.load("population.households").reset_index()
+    # )
+    # for pop in populations:
+    #     for (state, puma) in pop[["state", "puma"]].drop_duplicates().values:
+    #         assert puma in state_puma_map[state]
+
+
+@pytest.mark.parametrize(
+    "unit_id_col, address_id_col",
+    [
+        ("employer_id", "business_details.employer_address_id"),
+        ("household_id", "household_details.address_id"),
+    ],
+)
+@pytest.mark.skip(reason="Wait for state/puma to be added (MIC-3728)")
+def test_addresses_during_moves(simulants_on_adjacent_timesteps, unit_id_col, address_id_col):
+    """Check that unit (household and business) address details change after a move."""
+    other_address_cols = []  # TODO: add state/puma
+    address_cols = [address_id_col] + other_address_cols
+    for before, after in simulants_on_adjacent_timesteps:
+        # get the unique unit (household or employer) dataset for before and after
+        before_units = before.groupby(unit_id_col)[address_cols].first()
+        after_units = after.groupby(unit_id_col)[address_cols].first()
+        # reindex in case there are unique units to one of the datasets
+        total_index = before_units.index.union(after_units.index)
```

Comment:
I agree that intersection is correct rather than union. The `assert_frame_equal` below works because it operates only on households that have the same `address_id` in both `after` and `before` (and therefore must exist in both).